doc/feature/compartmentReport.md

```diff
@@ -0,0 +1,422 @@
+         *
+         * @param startTime the start time of the report
+         * @param endTime the end time of the report
+         * @param timestep the timestep between report frames
+         * @param dunit the unit of the data, e.g. mV
+         * @param tunit the unit of the time, e.g. ms
+         * @throw std::invalid_argument if any passed argument is invalid
+         */
+        void writeHeader( const float startTime, const float endTime,
+                          const float timestep, const std::string& dunit,
+                          const std::string& tunit );
+
+        /** Write the compartment counts for each section for one cell.
+         *
+         * Preconditions:
+         * - The report was open in write mode.
+         * - writeHeader has already been called.
+         *
+         * @param gid the GID of the cell
+         * @param counts the number of compartments per section
+         * @throw std::runtime_error if writing was not successful.
+         */
+        void writeMapping( const uint32_t gid,
+                           const uint16_ts& counts );
+
+        /** Write the compartment counts for each section for one cell.
+         *
+         * Preconditions:
+         * - The report was open in write mode.
+         * - writeMapping has already been called.
+         *
+         * @param gid the GID of the cell
+         * @param voltages the voltages per compartment to save
+         * @throw std::runtime_error if writing was not successful.
+         */
+        void writeCell( const uint32_t gid, const floats& voltages );
+
+        /** Finish writing the current frame and advance to the next one.
+         *
+         * Upon return getCurrentTime() has been increased to the nearest
+         * multiple of the timestep greater than the value before.
+         */
+        void endFrame();
+    }
+
+## Implementation
+
+This section will cover some implementation details related to the stream
+case as the file based IO is straight forward.
+
+The most two important decisions taken are:
+- The granularity of the messages sent won't be the whole frame but a cell.
+- The mapping will be always included as part of the frame.
+
+The rationale for these decisions follows.
+
+The current API for writing is already cell oriented, while this is convenient
+to write on memory mapped files or H5 files without requiring the client to
+assemble the whole buffer before writing, it's not consistent with the fact that
+reads are whole frame oriented. This means that in the stream case we have to
+assemble the frame at the writer or reader part of the implementation, or
+change the API and force the client to do it on the writer side. Requiring
+the client to assemble the frame is the easiest solution for the implementation
+of the stream based case but it requires bigger changes to adapt existing code
+both in the client side and in the implementation of existing plugins.
```

Comment:
client side is confusing. If I understand correctly, the assembly happens at the receiver end, but within the plugin?